File: backend/app/schemas/memory.py
"""Long-term memory audit schemas (Phase 11)."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel


class MemoryOut(BaseModel):
    id: uuid.UUID
    kind: str
    content: str
    confidence: float
    source_thread_id: uuid.UUID | None
    created_at: datetime
    # `None` means active. A caller does not need a separate `is_active` flag — the
    # audit UI's own job is showing *when* something stopped being active, not hiding
    # that fact behind a boolean.
    superseded_at: datetime | None


__all__ = ["MemoryOut"]

File: frontend/src/features/memory/api/memory.ts
/**
 * Long-term memory audit API contract (Phase 11).
 *
 * Schema mirrors `backend/app/schemas/memory.py`.
 */
import { z } from 'zod';

import { apiClient } from '@/shared/lib/api-client';

export const memoryKindSchema = z.enum(['preference', 'goal', 'recurring_finding']);
export type MemoryKind = z.infer<typeof memoryKindSchema>;

export const memorySchema = z.object({
  id: z.string(),
  kind: memoryKindSchema,
  content: z.string(),
  confidence: z.number(),
  source_thread_id: z.string().nullable(),
  created_at: z.string(),
  // `null` means active — see `MemoryOut`'s own docstring for why there is no separate
  // boolean flag.
  superseded_at: z.string().nullable(),
});
export type Memory = z.infer<typeof memorySchema>;

function withProfile(path: string, profileId?: string): string {
  return profileId ? `${path}?profile_id=${profileId}` : path;
}

export function listMemories(profileId?: string, signal?: AbortSignal): Promise<Memory[]> {
  return apiClient.get(withProfile('/api/v1/memory', profileId), z.array(memorySchema), signal);
}

export function deleteMemory(memoryId: string, profileId?: string): Promise<void> {
  return apiClient.delete(withProfile(`/api/v1/memory/${memoryId}`, profileId));
}
